auth_server/src/main.rs

```diff
@@ -0,0 +1,217 @@
+use auth_server::{
+    is_valid_credentials, AuthCode, Client, Credentials, IntrospectResponse, LoginForm,
+    LoginQueryParams, OauthTokenParams, Profile, Token, TokenParams, TokenResponse, TokenStatus,
+    User, MUTABLE_STORAGE,
+};
+use chrono::{DateTime, Duration, Utc};
+use url::Url;
+
+use actix_web::{
+    get, http::header::ContentType, post, web, App, HttpResponse, HttpServer, Responder,
+};
+
+#[get("/authorize")]
+async fn login(query: web::Query<LoginQueryParams>) -> impl Responder {
+    let LoginQueryParams {
+        client_id,
+        redirect_uri,
+        response_type,
+        scope,
+    } = query.into_inner();
+
+    let client = MUTABLE_STORAGE
+        .lock()
+        .unwrap()
+        .clients
+        .get(&client_id[..])
+        .cloned();
+
+    match client {
+        Some(cl) => HttpResponse::Ok()
+            .content_type(ContentType::html())
+            .body(format!(
+                "<h2>{} is requesting access to:</h2>
+                <form method=\"post\" action=\"/authorize\">
+                    <input type=\"hidden\" name=\"client_id\" value=\"{}\" />
+                    <input type=\"hidden\" name=\"redirect_uri\" value=\"{}\" />
+                    <input type=\"hidden\" name=\"response_type\" value=\"{}\" />
+                    <input type=\"hidden\" name=\"scope\" value=\"{}\" />
+                    <input type=\"text\" name=\"username\" placeholder=\"Username\" /><br>
+                    <input type=\"password\" name=\"password\" placeholder=\"Password\" /><br>
+                    <button type=\"submit\">Authorize</button>
+            </form>",
+                cl.name, client_id, redirect_uri, response_type, scope
+            )),
+        None => HttpResponse::BadRequest().body("Invalid client"),
+    }
+}
+
+#[post("/authorize")]
+async fn login_post(form: web::Form<LoginForm>) -> impl Responder {
+    let form = form.into_inner();
+    let creds = Credentials {
+        client_id: form.client_id.clone(),
+        client_secret: None,
+    };
+    let is_valid = is_valid_credentials(&creds);
+    if !is_valid {
+        return HttpResponse::Unauthorized().body("Invalid credentials");
+    }
+
+    let mut storage = MUTABLE_STORAGE.lock().unwrap();
```

Comment:
It would be beneficial to reduce lock times. One approach is to create a repository structure and implement storage operations directly within its methods.